### PR TITLE
[Examples] add a new example of tabby

### PR DESCRIPTION
**Pull request by sunny0826**
Add a Tabby example to guide users in quickly deploying Tabby on any cloud using SkyPilot.

### CODE REVIEW

File: llm/tabby/tabbyml.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+run: |
+  set -ex
+
+  mkdir -p tabby
+  cp config.toml tabby/
+
+  docker-compose down
+
+  if nvidia-smi; then
```
Comment (cblmemo) [inline llm/tabby/tabbyml.yaml:29]: redirect too?

File: llm/tabby/README.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
...//github.com/skypilot-org/skypilot) is an open-source framework from UC Berkeley for seamlessly running machine learning on any cloud. With a simple CLI, users can easily launch many clusters and jobs, while substantially lowering their cloud bills. Currently, [AWS](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#aws), [GCP](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#gcp),  [Azure](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#azure), [Lambda Cloud](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#lambda-cloud), [IBM](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#ibm), [Oracle Cloud Infrastructure (OCI)](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#oracle-cloud-infrastructure-oci), [Cloudflare R2](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#cloudflare-r2) and [Samsung Cloud Platform (SCP)](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#samsung-cloud-platform-scp) are supported. See [docs](https://skypilot.readthedocs.io/en/latest/) to learn more.
```
Comment (cblmemo) [inline llm/tabby/README.md:17]:
```suggestion
[**SkyPilot**](https://github.com/skypilot-org/skypilot) is an open-source framework from UC Berkeley for seamlessly running machine learning on any cloud. With a simple CLI, users can easily launch many clusters and jobs, while substantially lowering their cloud bills. Currently, [AWS](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#aws), [GCP](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#gcp), [Azure](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#azure), [Lambda Cloud](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#lambda-cloud), [IBM](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#ibm), [Oracle Cloud Infrastructure (OCI)](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#oracle-cloud-infrastructure-oci), [Cloudflare R2](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#cloudflare-r2) and [Samsung Cloud Platform (SCP)](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#samsung-cloud-platform-scp) are supported. See [docs](https://skypilot.readthedocs.io/en/latest/) to learn more.
```
(there are two consecutive space before [Azure])

**Review by cblmemo — COMMENTED** [review]
Thanks for the exciting demo!!! 🚀 Tried the YAML and it works great 💪🏻. Left several nits ;)
Also, a discussion: There is a lot of introduction content for the skypilot repository, which seems a little bit redundant inside the skypilot repo. Maybe we could remove some of them? cc @Michaelvll for a look 👀

**Comment by wsxiaoys** [comment]
nit: example of tabbyml -> tabby

Thanks for the PR! ❤️

**Comment by sunny0826** [comment]
@cblmemo Sorry for the somewhat late reply, I've been attending KubeCon China for the last few days. I fixed and tested most of what was in the review feedback. 

Unfortunately, I found a bug in my testing: when I opened the port with `ports: 8080`, the record 8080 was not added to the network security group created by skypilot, which prevented me from accessing Tabby via public IP. After I manually added port 8080, Tabby was able to access it properly.

**Additional context**

- cloud: Azure
- region australiaeast 

image

**Comment by cblmemo** [comment]
> @cblmemo Sorry for the somewhat late reply, I've been attending KubeCon China for the last few days. I fixed and tested most of what was in the review feedback.
> 
> Unfortunately, I found a bug in my testing: when I opened the port with `ports: 8080`, the record 8080 was not added to the network security group created by skypilot, which prevented me from accessing Tabby via public IP. After I manually added port 8080, Tabby was able to access it properly.
> 
> I haven't read the source code, so I'm not sure if this phenomenon is unique to Azure or if all clouds have similar issues.
> 
> **Additional context**
> 
> * cloud: Azure
> * region australiaeast
> 
> image

Thanks for reporting that! I tried this YAML with `accelerators: V100:1` commented out (due to quota issues) several times, and one of them failed. It seems like an occasional happened bug. Will dive deep into it 🫡

File: llm/tabby/tabby/config.toml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+[[repositories]]
```
Comment (Michaelvll) [inline llm/tabby/tabby/config.toml:1]: Just curious, what is this file used for? I don't see a place where we are using SkyPilot in the docker container. : )

**Review by Michaelvll — APPROVED** [review]
Thanks for adding this tabby example @sunny0826! This is fantastic! Left a question for the `toml` file. Otherwise, LGTM.

File: llm/tabby/tabby/config.toml (hunk 3 above)
Comment (wsxiaoys) [reply llm/tabby/tabby/config.toml:1]: tabby possesses the capability to execute fetching and parsing operations on these repositories, potentially utilizing them for augmented prompting. However, these features remain either undocumented or are accessible only through specific internal flags.

Comment (sunny0826) [reply llm/tabby/tabby/config.toml:1]: Thanks @wsxiaoys  for adding this, I was actually thinking about how to illustrate this.

File: llm/tabby/docker-compose.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+services:
+  scheduler:
+    image: tabbyml/tabby
+    container_name: tabby-scheduler
+    platform: linux/amd64
+    command: scheduler --now
+    volumes:
+      - "./tabby:/data"
+
```
Comment (wsxiaoys) [inline llm/tabby/docker-compose.yaml:11]:
```suggestion
```

Remove the scheduler part as it's not being officially documented yet

File: llm/tabby/docker-compose.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  tabby:
+    depends_on:
+      - scheduler
```
Comment (wsxiaoys) [inline llm/tabby/docker-compose.yaml:14]:
```suggestion
```
ditto

File: llm/tabby/docker-compose.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    restart: always
+    container_name: tabby
+    image: tabbyml/tabby
+    command: serve --model TabbyML/SantaCoder-1B
```
Comment (wsxiaoys) [inline llm/tabby/docker-compose.yaml:18]:
```suggestion
    command: serve --model TabbyML/StarCoder-1B
```

File: llm/tabby/tabby.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
+resources:
+  # you could change this section to any resources you want, like a CPU VM
+  accelerators: V100:1
```
Comment (wsxiaoys) [inline llm/tabby/tabby.yaml:3]:
```suggestion
  accelerators: T4:1
```

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: llm/tabby/tabby.yaml (hunk 7 above)
Comment (sunny0826) [reply llm/tabby/tabby.yaml]: done

**Comment by Michaelvll** [comment]
Thanks for adding this example @sunny0826 @wsxiaoys! I am going to merge it soon. : )

**Pull request merged**